sources/libClangSharp/ClangSharp.cpp

```diff
@@ -4701,13 +4701,16 @@ int64_t clangsharp_Cursor_getVtblIdx(CXCursor C) {
             if (VTableContextBase::hasVtableSlot(CMD)) {
                 VTableContextBase* VTC = getASTUnit(getCursorTU(C))->getASTContext().getVTableContext();
 
+                const CXXDestructorDecl* dCMD = dyn_cast<CXXDestructorDecl>(CMD);
+                const GlobalDecl globalDecl = dCMD ? GlobalDecl(dCMD, Dtor_Deleting) : GlobalDecl(CMD);
```

Comment:
Looks like this also needs handling for constructors. `CUDAGlobalAttr` looks to be covered implicitly.

I also don't think this is quite the right fix. I think it would be better for this to be extended to also accept `GlobalDecl` directly and to explicitly have `getDtorVtblIdx` and `getCtorVtblIdx` that users should call so they can override which `DtorType`/`CtorType` they attempt to get the VTBL index for, given they may not always want the deleting dtor.

This makes it more extensible for the future while also allowing something that "works by default".

Reply:
> Looks like this also needs handling for constructors…

Pardon, can't understand that. This block handles virtual functions, right? There can't be constructors here.

Reply:
Ah right, the `hasVtableSlot` check will prevent this from being called for constructors. Nevermind on that part.

Reply:
OK, thanks! Now I'm puzzled with the next one: virtual destructor declaration always creates the deleting destructor. What destructor type user can possibly select here?

Reply:
Its been a while since I've had to read the rules here, but as far as I remember:

Declaring a virtual destructor will cause a pair of entries to be created, a complete and a deleting version. The complete version is simply the destruction logic (that is `~MyClass()`) while the deleting version is a specially generated version that also calls the relevant `delete` operator after invoking the destruction logic. This means that there are actually 2 virtual entries that may exist and which is called can be dependent on the scenario and whether calling `delete` is correct or not.

Its entirely possible I'm misremembering and conflating the concepts with how the VTBL is actually laid out however.

Reply:
From https://itanium-cxx-abi.github.io/cxx-abi/abi.html#vtable
> The entries for virtual destructors are actually pairs of entries. The first destructor, called the complete object destructor, performs the destruction without calling delete() on the object. The second destructor, called the deleting destructor, calls delete() after destroying the object. Both destroy any virtual bases; a separate, non-virtual function, called the base object destructor, performs destruction of the object but not its virtual base subobjects, and does not call delete().

Reply:
The written above is correct as far as I am aware. Since the destructor is mapped to to `Dispose()` and it makes no sense to call the non-deleting version from outside (if that `Dispose()` needs to be called at all), I again can't understand why the other one should be exposed.

Reply:
ClangSharp is multiple parts, only one of which supports C#. It also is a general purpose wrapper over Clang itself, can expose the metadata for bindings via XML, and is used by the `win32metadata` project which itself is used for generating bindings for other languages, like Rust or even "modern C++" wrappers.

A power user may want to use this metadata to generate bindings for another language or even to synthesize their own types that "inherit" from a C++ type, in which case they may want to call the complete (non-deleting) destructor and call delete themselves.

Reply:
Attempted to implement what you asked for. Take a look, please.

Reply:
LGTM minus the two nits/fixes requested below.